src/sb3-serialize.ts

```diff
@@ -0,0 +1,219 @@
+            block.opcode = 'data_listcontents';
+            block.fields['LIST'] = {
+                name: 'LIST',
+                value: data[1],
+                id: data[2],
+                variableType: ScratchVariableTypes.LIST_TYPE,
+            };
+            if (data.length > 3) {
+                block.topLevel = true;
+                block.x = data[3];
+                block.y = data[4];
+            }
+            break;
+
+        default:
+            throw new Error(`sb-util: Unknown serialized block type ${type}`);
+    }
+    return block;
+}
+
+function deserializeFields(serialized: SB3SerializedFields): BlockFields {
+    const fields: BlockFields = {};
+    for (const [name, data] of Object.entries(serialized)) {
+        if (Array.isArray(data)) {
+            fields[name] = {
+                name,
+                value: data[0],
+                id: data[1],
+            };
+
+            if (name === 'BROADCAST_OPTION') {
+                fields[name].variableType = ScratchVariableTypes.BROADCAST_MESSAGE_TYPE;
+            } else if (name === 'VARIABLE') {
+                fields[name].variableType = ScratchVariableTypes.SCALAR_TYPE;
+            } else if (name === 'LIST') {
+                fields[name].variableType = ScratchVariableTypes.LIST_TYPE;
+            }
+        } else {
+            fields[name] = data;
+        }
+    }
+    return fields;
+}
+
+function deserializeInputs(
+    serialized: SB3SerializedInputs,
+    parentId: string,
+): { inputs: BlockProperties['inputs']; newBlocks: BlockProperties[] } {
+    const inputs: BlockProperties['inputs'] = {};
+    const newBlocks: BlockProperties[] = [];
+
+    function parse(data: SB3SerializedBlockArray | string, shadow: boolean = false): string {
+        if (typeof data === 'string') {
+            return data;
+        }
+        const parsed = deserializeBlockArray(data, newBlockId(), parentId, shadow);
+        newBlocks.push(parsed);
+        return parsed.id;
+    }
+
+    for (const [name, data] of Object.entries(serialized)) {
+        if (Array.isArray(data)) {
+            const [type, blockData, shadowData] = data;
+            let block = null;
+            let shadow = null;
+            if (type === SB3SerializedInputType.INPUT_SAME_BLOCK_SHADOW) {
+                block = parse(blockData, true);
+                shadow = block;
+            } else if (type === SB3SerializedInputType.INPUT_BLOCK_NO_SHADOW) {
+                block = parse(blockData);
+            } else if (type === SB3SerializedInputType.INPUT_DIFF_BLOCK_SHADOW) {
+                block = parse(blockData);
+                shadow = parse(shadowData, true);
+            }
+            inputs[name] = { name, block, shadow };
+        } else {
+            inputs[name] = data as BlockInput;
+        }
+    }
+
+    return { inputs, newBlocks };
+}
+
+function deserializeBlockObject(serialized: SB3SerializedBlockObject, id: string): BlockProperties[] {
```

Comment:
oh yeah - one block "object" in sb3 land can yield multiple blocks!  how interesting!